### ISSUE TITLE
Next 13 appDir: next/script with 'beforeInteractive' executing too late

### ISSUE THREAD

**Issue by rafma0** [vercel/next.js#43402]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 18.4.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.0.6-canary.1
      eslint-config-next: 13.0.6-canary.1
      react: 18.2.0
      react-dom: 18.2.0

### Describe the Bug

I'm trying to execute a simple script as early as possible to set a `data-theme` in my html, and I use it to get the colors for the theme from the .css. Using `next/script` with `beforeInteractive` in `app/layout`, it takes too long to execute and we can see the page "flashing"

Also, if I move the `<Script>` from the `<body>` to the `<html>` like it shows in the documentation, I get hydration errors
```
Warning: Cannot render a sync or defer <script> outside the main document without knowing its order. Try adding async="" or moving it into the root <head> tag.`
Warning: Expected server HTML to contain a matching <script> in <html>.
Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
```
(but maybe this last part should be in another issue?)

### Expected Behavior

To be honest I'm not sure this is a bug but I'm assuming it is since it works as intended if I use `<script>` instead. The expected behavior is having this script execute before any element with wrong colors is shown on any page.

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/rafma0/next-beforeinteractive

### To Reproduce

Load the root page from the repro link and you will see the page flashing to change the font and background colors.

Comment the `<Script>` in `<body>` and uncomment the `<script>` in `<head>` and the page wont't blink anymore (but in dev mode you get a warning `Warning: Prop 'data-theme' did not match. Server: "dark" Client: "light"`

**Comment by rnantes** [comment]
Having the same issue. This simple example does not work, and will cause a flash on reload.

``` javascript
export default function RootLayout({ children, } : {  children: React.ReactNode; }) {
return (
    <html lang="en" >
            <body>
                <div> hello world </div>
                <Script strategy="afterInteractive" id="dark-theme"
                      dangerouslySetInnerHTML={{
                          __html: `document.querySelector('html').classList.add('dark-theme');`,
                      }}
                />
          </body>
     </html>
);
}
``` 

Beta Layout Script docs:
https://beta.nextjs.org/docs/optimizing/scripts#layout-scripts

**Comment by tmrp** [comment]
What solved the flashing issue for me was using the standard `<script>` tag with a `defer` attribute:
 
```js
<script defer id="dark-theme"
  dangerouslySetInnerHTML={{
    __html: `document.querySelector('html').classList.add('dark-theme');`,
 }}
></script>
```

I know it's not directly related to the bug, but this might unblock/be a workaround.

**Comment by itsjavi** [comment]
This is still an issue on next 13.3.0.

With @tmrp solution it works, even without defer, but it's less than ideal. The script is added at the end of the head, after the CSS, so there is a small chance of flickering.

It also causes an error because of a server/client mismatch in the className

**Comment by rafma0** [comment]
It's been 9 months since I opened this, I've been using the `<script>` in the `<head>` as workaround

I just updated all packages in the the repro ex:  next `13.0.8-canary.0` -> `13.4.14-canary.2` etc and the behavior is exact the same

**Comment by AshConnolly** [comment]
For others having a similar issue @rauchg himself is using the script tag solution for his theming - https://rauchg.com/. 

See script implementation here - https://github.com/rauchg/blog/blob/main/app/layout.tsx#L45 and check out theme-effect.ts too, its a really elegant solution 👍

**Comment by rafma0** [comment]
Not directly related to the bug, but let me share how I ended up doing themes in a new project.

to store the current theme, good old cookie

```tsx
import { cookies } from 'next/headers'

export default function RootLayout ({ children }: Props) {
  const cookieStore = cookies()
  const theme = cookieStore.get('theme')?.value

  return (
    <html lang='en' data-theme={theme || 'auto'}>
      <body>{children}</body>
    </html>
  )
}
```

then I have a client component to toggle themes
```ts
const toggleTheme = () => {
  const currentTheme = document.documentElement.getAttribute('data-theme') || ''

  const themes = ['light', 'dark', 'green', 'auto']
  const index = themes.indexOf(currentTheme)
  const newTheme = themes[(index + 1) % themes.length]

  document.documentElement.setAttribute('data-theme', newTheme)
  Cookie.set('theme', newTheme, { expires: 365 }) // js-cookie
}
```

and my themes.scss (imported in globals.scss)
```scss
[data-theme=green] {
  --text: rgb(0, 190, 0);
  --background: rgb(0, 40, 0);
}

@mixin light($name) {
  [data-theme=#{$name}] {
    --text: #333;
    --background: #fbfbfb;
  }
}
@mixin dark($name) {
  [data-theme=#{$name}] {
    --text: #eee;
    --background: #101010;
  }
}

@include light('light');
@include dark('dark');

@media (prefers-color-scheme: light) {
  @include light('auto');
}
@media (prefers-color-scheme: dark) {
  @include dark('auto');
}
```

This is similar to what `github.com` does for multiple themes. You have to repeat one of your themes for the 'auto' variants but it works without any js on load.

I still think using `next/script` with `beforeInteractive` should work without blinking or needing to suppress hydration warning based on the docs https://nextjs.org/docs/app/api-reference/components/script#beforeinteractive but at least we have options. I'll probably go with the cookie solution when possible.
